### PR TITLE
Add OpenID Connect Support

### PR DESCRIPTION
**Pull request by adamvy**

### CODE REVIEW

File: src/foam/nanos/auth/jwks.jrl [hunk 1]
```diff
@@ -0,0 +1,26 @@
+// imported from https://www.googleapis.com/oauth2/v3/certs from https://accounts.google.com/.well-known/openid-configuration
+// on August 19 2024
+// TODO: Fetch these dynamically as needed
+
+p({
```
Comment (adamvy) [inline src/foam/nanos/auth/jwks.jrl]: fetch this with a daily cron?

File: src/foam/nanos/auth/oidc/OIDCWebAgent.java [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+import java.net.URL;
+import java.net.URLEncoder;
+import java.nio.charset.StandardCharsets;
+
+import org.apache.commons.io.IOUtils;
+import foam.nanos.logger.Logger;
+import foam.nanos.google.api.auth.GoogleApiCredentials;
+
+public class OIDCWebAgent implements WebAgent {
+    private static final String TOKEN_ENDPOINT = "https://oauth2.googleapis.com/token";
```
Comment (adamvy) [inline src/foam/nanos/auth/oidc/OIDCWebAgent.java]: inject an openid provider with all the openid specifics

File: src/foam/nanos/auth/AuthenticationService.js [hunk 3]
```diff
@@ -0,0 +1,83 @@
+foam.CLASS({
+    package: 'foam.nanos.auth',
+    name: 'AuthenticationService',
```
Comment (adamvy) [inline src/foam/nanos/auth/AuthenticationService.js]: rename this to login service

File: src/foam/nanos/u2/navigation/SignIn.js [hunk 4]
```diff
@@ -275,6 +277,55 @@ foam.CLASS({
...
+    {
+      name: 'signInWithGoogle',
+      label: 'Sign in With Google',
+      section: 'footerSection',
+      isAvailable: function(showAction) { return showAction },
+      code: async function(X) {
+        // TODO: Validate nonce
+        var nonce = crypto.randomUUID();
+
+        var reqParams = {
```
Comment (adamvy) [inline src/foam/nanos/u2/navigation/SignIn.js]: add a comment showing how to do local redirect version

File: src/foam/nanos/auth/services.jrl [hunk 5]
```diff
@@ -194,9 +207,11 @@ p({
   "name":"languageDAO",
   "lazy":false,
   "serve":true,
+  "authenticate": false,
```
Comment (adamvy) [inline src/foam/nanos/auth/services.jrl:207]: remove authenticate: false

File: src/foam/nanos/auth/Credentials.js [hunk 6]
```diff
@@ -0,0 +1,25 @@
+foam.CLASS({
```
Comment (adamvy) [inline src/foam/nanos/auth/Credentials.js]: split to multiple files

File: src/foam/typescript/model/TypeScript.js [hunk 7]
```diff
@@ -0,0 +1,74 @@
+foam.CLASS({
```
Comment (adamvy) [inline src/foam/typescript/model/TypeScript.js:1]: remove from PR

File: src/foam/typescript/TypeScriptModelTransformer.js [hunk 8]
```diff
@@ -0,0 +1,76 @@
+foam.CLASS({
```
Comment (adamvy) [inline src/foam/typescript/TypeScriptModelTransformer.js:1]: remove from PR

File: src/foam/nanos/auth/services.jrl (hunk 5 above)
Comment (adamvy) [reply src/foam/nanos/auth/services.jrl:207]: @jlhughes does this need to be removed?  countryDAO is also authenticate: false

File: src/foam/typescript/model/TypeScript.js (hunk 7 above)
Comment (adamvy) [reply src/foam/typescript/model/TypeScript.js:1]: @jlhughes I'm just going to leave these here if that's okay with you

File: src/foam/nanos/auth/User.js [hunk 9]
```diff
@@ -978,24 +978,23 @@ foam.CLASS({
...
-        Session session = x.get(Session.class);
-        if ( session == null ) {
-          throw new AuthenticationException("No session exists.");
+        
+        // check if user login enabled
+        if ( ! getLoginEnabled() ) {
+          throw new AccessDeniedException();
         }
-
-        if ( this instanceof LifecycleAware && ((LifecycleAware) this).getLifecycleState() != LifecycleState.ACTIVE ) {
```
Comment (adamvy) [inline src/foam/nanos/auth/User.js:997]: this is redundant with the check a few lines above I believe so i removed it

File: src/foam/nanos/auth/UserAndGroupAuthService.js [hunk 10]
```diff
@@ -124,98 +132,22 @@ foam.CLASS({
...
-      ],
-      javaThrows: ['foam.nanos.auth.AuthenticationException'],
-      javaCode: `
-      try {
-        if ( user == null ) {
-          throw new UserNotFoundException();
-        }
-        user.validateAuth(x);
-        // check if user enabled
-        if ( user.getLifecycleState() != foam.nanos.auth.LifecycleState.ACTIVE ) {
```
Comment (adamvy) [inline src/foam/nanos/auth/UserAndGroupAuthService.js:153]: lifecycel state is checked in the above user.validateAuth

File: src/foam/nanos/auth/UserAndGroupAuthService.js [hunk 11]
```diff
@@ -124,98 +132,22 @@ foam.CLASS({
...
-        // check if group enabled
-        X userX = x.put("subject", new Subject.Builder(x).setUser(user).build());
-        Group group = user.findGroup(userX);
-        if ( group != null && ! group.getEnabled() ) {
-          throw new AccessDeniedException();
-        }
-        if ( ! Password.verify(password, user.getPassword()) ) {
-          throw new InvalidPasswordException();
-        }
-        if ( ! user.getEmailVerified() ) {
```
Comment (adamvy) [inline src/foam/nanos/auth/UserAndGroupAuthService.js:169]: verified email check was moved to user.validateAuth

File: src/foam/nanos/auth/UniqueUserService.js [hunk 12]
```diff
@@ -15,7 +15,7 @@ foam.INTERFACE({
       name: 'getUser',
       type: 'foam.nanos.auth.User',
       documentation: 'Helper logic function to reduce code duplication',
-      args: [ 'Context x', 'String identifier', 'String password' ]
```
Comment (adamvy) [inline src/foam/nanos/auth/UniqueUserService.js:18]: password was unused

File: src/foam/nanos/auth/UserAndGroupAuthService.js [hunk 13]
```diff
@@ -124,98 +132,22 @@ foam.CLASS({
         return subject;
       `
     },
-    {
-      name: 'loginHelper',
```
Comment (adamvy) [inline src/foam/nanos/auth/UserAndGroupAuthService.js:128]: this code was moved to LoginService so that OIDCWebAgent can access it

**Comment by adamvy** [comment]
@jlhughes this should be ready for review now.  Ended up removing the JWTAuthService, as the OIDCWebAgent can be used to just redirect the current page instead of opening an auth window.

Added some comments here and there to explain what I changed.  Thanks for the review!

File: src/foam/nanos/auth/services.jrl (hunk 5 above)
Comment (jlhughes) [reply src/foam/nanos/auth/services.jrl:207]: @adamvy languageDAO and countryDAO nspecs should be 
authenticate: false ( when true, the user must have service._nspec-id_ permission)
along with 
EasyDOA..
setAuthorize(true)  

EasyDAO with authorize: true will use the GlobalReadAuthorizer (all can read, permission required to update), which is what we want. 
The setAuthorizer is optional if the prefix  that was used for permissions was not the 'of' in lowercase. 
So for both language and country models, neither require the explicit setAuthorizer(...). 

But they are both correct as they stand.

**Review by jlhughes — APPROVED** [review]

**Comment by jlhughes** [comment]
@adamvy we are ready on the nanopay side.  Let me know when you are ready and I'll merge everything together.

File: src/foam/nanos/u2/navigation/SignIn.js [hunk 14]
```diff
@@ -182,6 +185,60 @@ foam.CLASS({
...
+    {
+      name: 'signInWithOIDC',
+      code: async function(provider) {
+        // TODO: Validate nonce
+        var nonce = crypto.randomUUID();
+
+        var reqParams = {
+          response_type: 'code',
+          client_id: provider.clientId,
+          scope: 'openid email',
```
Comment (adamfox-nanopay) [inline src/foam/nanos/u2/navigation/SignIn.js:198]: not sure if its necessary for this pr, but would be good if this was configurable somehow ... e.g for intuit openid we also used scope for fullname, phone number)

Comment (adamfox-nanopay) [reply src/foam/nanos/u2/navigation/SignIn.js:198]: or maybe it is not necessary at all for just sign in

**Review by adamfox-nanopay — APPROVED** [review]
lgtm

**Pull request merged**